### ISSUE TITLE
warning "property never defined" disappears when unused class is required

### ISSUE THREAD

**Issue by myphysicslab** [google/closure-compiler#826]
The warning `property {0} never defined on {1}` disappears when an unused class is `goog.require`d.

I have been noticing several cases where the compiler was failing to provide warnings on code where warnings should occur; this is the first case I've been able to track down and reduce to a simplified case. I am losing my trust in the compiler!

I created a simplified example which involves 4 interfaces and 2 concrete classes. It can be found compressed at http://www.myphysicslab.com/rmnp/test20150216.zip or uncompressed at http://www.myphysicslab.com/rmnp/test20150216

The example compiles the class called AbstractGraph. The problem occurs or not depending on whether this line is included in AbstractGraph:

    goog.require('myphysicslab.lab.VarsArrayStandard');

That class, VarsArrayStandard, is never used in AbstractGraph. So it is unexpected that requiring it changes the compiler behavior.

*Without* the goog.require(VarsArrayStandard) you get this warning (as you should):

```
./lab/AbstractGraph.js:28: WARNING - Property addObserver never defined on myphysicslab.lab.HasVariables
  this.hasVars_.addObserver(this);
                ^
```
*With* the goog.require(VarsArrayStandard) that **warning disappears** -- which is wrong.

(BTW, is there any way to make this an error instead of a warning?)

Note that when you use the compiler option `--jscomp_warning=reportUnknownTypes` then you see this warning in both cases:
```
./lab/AbstractGraph.js:28: WARNING - could not determine the type of this expression
  this.hasVars_.addObserver(this);
  ^
```
I have been unable to make that `reportUnknownTypes` warning disappear -- using casts does not help. (That `reportUnknownTypes` option is not generally useful because it emits thousands of warnings on closure library, as well as some other warnings on my code that I cannot make go away.)

Here is the version of the compiler I am using.

```
Closure Compiler (http://github.com/google/closure-compiler)
Version: v20150126-82-g85da5b6
Built on: 2015/02/11 20:55
```
Execute the shell script `compile.sh` to build the example. You will need to modify the locations of the closure-compiler and closure-library in that shell script.

Speculation
----------------
My guess is that the compiler conflates the interface HasVariables with the class VarsArrayStandard, because VarsArrayStandard is the only class that can possibly implement HasVariables. Then when the compiler decides whether it is OK to call `addObserver` on the HasVariables object, it instead asks whether it's OK to call `addObserve`r on a VarsArrayStandard object. Which is indeed OK.

This is a problem because if there were another class that implements HasVariables but does not implement Subject (which defines the addObserver method), then you would encounter a run-time error from passing that other class to AbstractGraph.

You might say "just don't require classes that are not used". But this came up because I was wanting to test if the HasVariables object was an instance of VarsArrayStandard, so that I could safely call addObserver on it. (I admit this is perhaps not the best OOP design.)

**Comment by concavelenz** [comment]
The "missing property" as it was originally written was a "can not possibly be defined check".   Any property that could possibly exist was allowed (any subclass property).  @interface declaration are loosely checked.   When @struct was introduced, it became possible to say "only allow the properties known to be part of this object" but this doesn't extend to @interface.   

The new type inference throws away the early assumptions that this was necessary, and @interfaces are strictly checked.  If you aren't using type based optimization (which haven't been updated yet) or if you can type check separately, the new type inference might work better for you.

**Comment by myphysicslab** [comment]
Indeed, after adding compile option `--new_type_inf` to the test example, the desired warning now appears.  However, so do 65 other warnings on things in closure-library.  This should be useful when closure-library is updated.

Can you explain more about these terms in your comment?
+ **type based optimization**
+ **type check separately**

(I suppose I'm not doing them if I don't know what they mean)

In the meantime, if I'm not using --new_type_inf I'll be more cautious when using interfaces.

**Comment by concavelenz** [comment]
@myphysicslab 
"type based optimizations" you can read more about here:
https://github.com/google/closure-compiler/wiki/Experimental-Type-Based-Property-Renaming

By "type check separately", I simply meant that you could use the new type inference for additional checking by running the compiler with it enabled and discarding the output and using the old type inference when optimizing.

**Comment by myphysicslab** [comment]
I'm satisfied that this is a situation which new type inference (NTI) handles correctly.  The limitations of the old compiler means it can probably not be fixed in the old compiler.
